drivers: sensor: bme68x_iaq: Add support for SPI bus

purpl3F0x opened the pull request:
Adds support for a bme680x sensor SPI bus.

NordicBuilder commented:
Thank you for your contribution!
It seems you are not a member of the nrfconnect GitHub organization. External contributions are handled as follows:
Large contributions, affecting multiple subsystems for example, may be rejected if they are complex, may introduce regressions due to lack of test coverage, or if they are not consistent with the architecture of nRF Connect SDK.
PRs will be run in our continuous integration (CI) test system.
If CI passes, PRs will be tagged for review and merged on successful completion of review. You may be asked to make some modifications to your contribution during review.
If CI fails, PRs may be rejected or may be tagged for review and rework.
PRs that become outdated due to other changes in the repository may be rejected or rework requested.
External contributions will be prioritized for review based on the relevance to current development efforts in nRF Connect SDK.  Bug fix PRs will be prioritized.
You may raise issues or ask for help from our Technical Support team by visiting https://devzone.nordicsemi.com/.

*Note: This comment is automatically posted and updated by the Contribs GitHub Action.*

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-13535
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

purpl3F0x commented:
Should I include a commit message body due avoid the compliance error ? title is pretty descriptive of the commit.

maxd-nordic commented:
> Should I include a commit message body due avoid the compliance error ? title is pretty descriptive of the commit.

Yes, the compliance rules require that. Sorry about the hassle! You could add something like "This patch adds SPI bus support for the bme68x_iaq driver."

purpl3F0x commented:
> Yes, the compliance rules require that. Sorry about the hassle! You could add something like "This patch adds SPI bus support for the bme68x_iaq driver."

Νo problem, wasn't sure if it will be a blocking matter. Hopefully it's okay now.

maxd-nordic commented on drivers/sensor/bme68x_iaq/bme68x_iaq.c:
```diff
@@ -111,18 +113,62 @@ static void state_save(const struct device *dev)
 	memcpy(&buf[1], reg_data_ptr, len);
 
-	return i2c_write_dt(&bme68x_i2c_spec, buf, ARRAY_SIZE(buf));
+	return i2c_write_dt(&bme68x_i2c_specc, buf, ARRAY_SIZE(buf));
```
```suggestion
	return i2c_write_dt(&bme68x_i2c_spec, buf, ARRAY_SIZE(buf));

```

purpl3F0x replied:
oops

maxd-nordic commented on drivers/sensor/bme68x_iaq/bme68x_iaq.c:
```diff
@@ -401,12 +450,20 @@ static int bme68x_bsec_init(const struct device *dev)
+	}
+	data->dev.intf = BME68X_SPI_INTF;
+#elif BME68x_BUS_I2C
+	if (!device_is_ready(bme68x_i2c_spec)) {
```
```suggestion
	if (!device_is_ready(bme68x_i2c_spec.bus)) {
```

maxd-nordic commented on drivers/sensor/bme68x_iaq/bme68x_iaq.h:
```diff
@@ -7,13 +7,34 @@
+#if BME68x_BUS_SPI
+	#include <zephyr/drivers/spi.h>
+#elif BME68x_BUS_I2C
+	#include <drivers/i2c.h>
```
```suggestion
	#include <zephyr/drivers/i2c.h>
```

maxd-nordic reviewed (changes requested):
With the suggested changes, the i2c build works again.

purpl3F0x commented:
Sorry for the typos.

Btw, somewhat unrelated to the commit, should support for other BSEC metrics (VOC, CO2 etc.) be added?

maxd-nordic commented:
> Sorry for the typos.
> 
> Btw, somewhat unrelated to the commit, should support for other BSEC metrics (VOC, CO2 etc.) be added?

Oh, if you know how to add these, this could be pretty cool! Last time I checked it looked a bit involved to get other metrics.

For now, it would be good to add a build test for the SPI case as well. For this, we would add the following file at `nrf/samples/sensor/bme68x_iaq/boards/nrf9160dk_nrf9160_ns.overlay`:

```
/*
 * Copyright (c) 2023 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
 */
&arduino_spi {
	status = "okay";

        bme680@0 {
                compatible = "bosch,bme680";
                reg = <0x0 >;
		spi-max-frequency = <DT_FREQ_M(8)>;
        };
};
```
And adding this to `nrf/samples/sensor/bme68x_iaq/sample.yaml`:
```
  sample.sensor.bme68x.spi:
    depends_on: spi
    harness: sensor
    integration_platforms:
      - nrf9160dk_nrf9160_ns
    platform_allow: >-
      nrf9160dk_nrf9160_ns
    tags: sensors
```

purpl3F0x commented:
> sample.sensor.bme68x.spi:
>     depends_on: spi
>     harness: sensor
>     integration_platforms:
>       - nrf9160dk_nrf9160_ns
>     platform_allow: >-
>       nrf9160dk_nrf9160_ns
>     tags: sensors

Added the suggested additions, I also renamed the first test to `sample.sensor.bme68x.i2c` to be more descriptive.

maxd-nordic commented on samples/sensor/bme68x_iaq/boards/nrf9160dk_nrf9160_ns.overlay:
```diff
@@ -0,0 +1,14 @@
+                reg = <0x0 >;
+		spi-max-frequency = <DT_FREQ_M(8)>;
+        };
+};
```
```suggestion
};

```

maxd-nordic commented on samples/sensor/bme68x_iaq/sample.yaml:
```diff
@@ -35,3 +35,11 @@ tests:
       ordered: true
       regex:
         - "(\\d+\\.\\d+); press: (\\d+\\.\\d+); humidity: (\\d+\\.\\d+); iaq: (\\d+)"
+    sample.sensor.bme68x.spi:
```
```suggestion
  sample.sensor.bme68x.spi:
```

maxd-nordic commented on samples/sensor/bme68x_iaq/sample.yaml:
```diff
@@ -35,3 +35,11 @@ tests:
+  sample.sensor.bme68x.spi:
+    depends_on: spi
+    harness: sensor
+    integration_platforms:
```
looks like we have to remove the `integration_platforms` part for the new test - my fault

maxd-nordic commented:
could you rebase?

purpl3F0x commented:
> could you rebase?

should I re-rebase now ? 😅

maxd-nordic commented:
@purpl3F0x Thanks for your PR! It will be included in the next release. :)
